source/common/network/lc_trie.h

```diff
@@ -0,0 +1,588 @@
+   *                    value for a sub-trie.
+   * @param root_branching_factor supplies the branching factor at the root. The paper suggests for
+   *                              large LC-Tries to use the value '16' for the root branching
+   *                              factor. It reduces the depth of the trie.
+   */
+  LcTrie(const std::vector<std::pair<std::string, std::vector<Address::CidrRange>>>& tag_data,
+         double fill_factor = 0.5, uint32_t root_branching_factor = 0);
+
+  /**
+   * Retrieve the tag associated with the CIDR range that contains `ip_address`. Both IPv4 and IPv6
+   * addresses are supported.
+   * TODO(ccaraman): Change to getTags and std::vector<std::string> when nested prefixes are
+   * supported.
+   * @param  ip_address supplies the IP address.
+   * @return tag from the CIDR range that contains 'ip_address'. An empty string is returned
+   *             if no prefix contains 'ip_address' or there is no data for the IP version of the
+   *             ip_address.
+   */
+  std::string getTag(const Network::Address::InstanceConstSharedPtr ip_address) const;
+
+private:
+  /**
+   * Extract n bits from input starting at position p.
+   * @param p supplies the position.
+   * @param n supplies the number of bits to extract.
+   * @param input supplies the IP address to extract bits from. The IP address is stored in host
+   *              byte order.
+   * @return extracted bits in the format of IpType.
+   */
+  template <class IpType, uint32_t address_size = CHAR_BIT * sizeof(IpType)>
+  static IpType extractBits(uint32_t p, uint32_t n, IpType input) {
+    // The IP's are stored in host byte order.
+    // By shifting the value to the left by p bits(and back), the bits between 0 and p-1 are zero'd
+    // out. Then to get the n bits, shift the IP back by the address_size minus the number of
+    // desired bits.
+    return input << p >> (address_size - n);
+  }
+
+  /**
+   * Removes n bits from input starting at 0.
+   * @param n supplies the number of bits to remove.
+   * @param input supplies the IP address to remove bits from. The IP address is stored in host
+   *              byte order.
+   * @return input with 0 through n-1 bits cleared.
+   */
+  template <class IpType, uint32_t address_size = CHAR_BIT * sizeof(IpType)>
+  static IpType removeBits(uint32_t n, IpType input) {
+    // The IP's are stored in host byte order.
+    // By shifting the value to the left by n bits and back, the bits between 0 and n-1
+    // (inclusively) are zero'd out.
+    return input << n >> n;
+  }
+
+  // IP addresses are stored in host byte order to simplify
+  typedef uint32_t Ipv4;
+  typedef absl::uint128 Ipv6;
+
+  // Helper methods to retrieve the string representation of the address in IpPrefix using
+  // inet_ntop. These strings are used in the nested prefixes exception messages.
+  // TODO(ccaraman): Remove once nested prefixes are supported.
+  static std::string toString(const Ipv4& input) {
```

Comment:
For these can you just construct an Ipv4Instance or Ipv6Instance and return the string from that? Seems simpler.

Reply:
Good idea. Fixed.